crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -125,11 +124,22 @@ pub trait SystemParamFetch<'world, 'state>: SystemParamState {
     unsafe fn get_param(
         state: &'state mut Self,
         system_meta: &SystemMeta,
-        world: &'world World,
+        world: &mut SemiSafeCell<'world, World>,
```

Comment:
Why do we pass `&mut SemiSafeCell` here? Isn't this like passing `&mut &T` ? Especially since we're not expecting the implementations here to change the pointer inside SemiSafeCell we should probably pass `SemiSafeCell` directly by-value here.

Reply:
I'm passing a `&mut` so that the system params of a system can all mutate the same cell (like `Option::take`). This way `get_param` will panic if a combination of params would be unsound.

 e.g. `Query<()>` and `&mut World`. `QueryState::get_param` turns the cell into a `Ref` so `WorldMutState::get_param` will panic. If you swap them around, `WorldMutState::get_param` turns the cell into `Empty`, so `QueryState::get_param` will panic.

Reply:
This feels like a half-measure to me tbh... especially given how `SemiSafeCell` is `Copy`, I feel like we shouldn't try to do this but rather just make sure such systems panic on startup - which they already should be doing.

Reply:
I'm all ears if you have a "full" measure in mind that makes systems panic on startup.

Reply:
I'm actually inclined to agree that we should be harsher here - we should split `system_param`'s `get_param` into two functions - one taking `&'_ mut Option<&'world mut World>` and one taking `&World`. Most base params would panic on one or the other, apart from `Local`

That also avoids the branching

Reply:
I'm not seeing the full picture. Can you step through how you'd plumb references to the params, starting from the `&mut World` passed into the executor? 

If we add a `get_exclusive_param`, where do we determine if we call that or `get_param` and how is that not a branch?

Reply:
So we determine that in the executor, based on the result of `needs_exclusive`.

That is, this is equivalent to our current version, except instead of creating a `SemiSafeCell::Mut`, we'd instead call `run_exclusive` on `System`, which itself would call `get_exclusive_param`. So it's a single branch per system, instead of `O(n)` branches in the number of system parameters (alternatively, `System` itself could unwrap the `SemiSafeCell`, if that were easier - the unsafety is at least constrained there, since there is only the one valid implementation of `System`).

Reply:
OK, I see now. But how should `get_exclusive_param` work with tuples, `ParamSet`, and params generated using the derive macro? We can't just unwrap the option and call `get_param`.

Reply:
`ParamSet` is a special case, but the other two are easy - each method just calls the same method on the 'sub' items.

`ParamSet` would effectively need to store an `Either<&mut World, &World>`, I think - if any of the parameters are exclusive it stores the `&mut World`, else the `&World` - then any non-exclusive parameters would be created by creating a shared reference from the exclusive reference.

Reply:
`get_exclusive_param` can simply panic for tuples and ones made by the derive macro, since it's impossible for these to be correct. `ParamSet` on the other hand would simply store a reference to the exclusive world inside itself, either with an enum or some form of untyped thing, then use it to get the actual param once one of the methods are called.

Reply:
> I'm all ears if you have a "full" measure in mind that makes systems panic on startup.

Conflicting queries already panic on startup no? I imagine the same checking that happens during the initialization phase would also cause panics for this.

Reply:
Hmm, I don't understand why it should panic.

I thought one of the motivations for this refactor is to allow `Locals` in exclusive systems? I'm surprised to learn we don't want that.

Reply:
`Local` s wouldn't panic since they don't register any accesses. However, similar to how e.g. multiple `ResMut` accessing the same resource will be seen as a conflict while the state for these is being initialized and accesses getting written to `SystemMeta`, I'm pretty sure having any other accesses alongside an exclusive one would cause a panic in that step.

Reply:
But you are suggesting to make any `&mut World` systems unconditionally panic when called, which is certainly a bold strategy. All systems currently have a `SystemParam` tuple, as a consequence of how `SystemParamFunction` (currently) works. You're saying we should panic if that tuple contains `&mut World`, thereby making any system with `&mut World` panic.

Reply:
This also made me notice that messages like this:

https://github.com/bevyengine/bevy/blob/7ecd69e8425e8269f3df16b8b5f56cb557fc5ce3/crates/bevy_ecs/src/system/system_param.rs#L462

should be change, since now it could also be triggered by other previously added accesses, whether that is a `ParamSet`, `&mut World` or a custom `SystemParam`. Not an issue introduced here, but something relevant to point out

Reply:
> You're saying we should panic if that tuple contains &mut World, thereby making any system with &mut World panic.

I suppose one-tuples would need to be exempted from this.

Reply:
I think this has been resolved. 

Using the new trait method, a system will panic when you call `IntoSystem::into_system` if its params can mix shared and exclusive world access (detecting incompatible component access still has to wait until `initialize`).

Trusting that as our soundness guarantee, `get_param` just takes a `MaybeUnsafeCell<World>` by value (it's `Copy`) and turns it into whatever the system needs. Main schedule systems will all semantically receive a `&'world UnsafeCell<World>` and the executor will ensure that exclusive systems never run in parallel.

Since there's no `Option::take`-like behavior, users can still cause UB by accidentally or deliberately saying an exclusive system param isn't exclusive, but the `SystemParamState` trait is marked as `unsafe`, so I believe that's acceptable.